ucp/_libs/ucp_py.pyx

```diff
@@ -43,17 +42,30 @@ include "ucp_py_buffer_helper.pyx"
 
 PENDING_MESSAGES = {}  # type: Dict[Message, Future]
 UCX_FILE_DESCRIPTOR = -1
-reader_added = 0
-UCP_INITIALIZED = False
 LOGGER = None
 listener_futures = set()
 
+def initialized():
+    global UCX_FILE_DESCRIPTOR
+    return UCX_FILE_DESCRIPTOR >= 0
+
+def update_pending_messages():
+    """Checking for finished futures in `PENDING_MESSAGES`"""
+    dones = []
+    for msg, fut in PENDING_MESSAGES.items():
+        assert(not fut.done())
```

Comment:
Are you always guaranteed that futures are which in `PENDING` are `not done` ?  Could a message in PENDING be done by the time the assertion check is claled ?

Reply:
Good point, fixed in f50ac3ced6e511c9952c298662e3f9a1d758acbd where done futures are passed through. Notice, calling `.set_result()`on a done future is illigal.

Reply:
If you're concerned about state changing within this for loop then that might not be a concern.  

Recall that because we're using only one thread, race conditions like this don't occur, unless there is an `await` statement somewhere, during which any other coroutine can come in and do anything.  This is one of the nice parts about using async/await and an event loop, rather than normal threads.

Reply:
Hmm, I ran [dask-cudf-full-test.py](https://github.com/quasiben/cudf-ucx-fail-scripts/blob/master/dask-cudf-full-test.py) and was able to generate an AssertionError in this code path.  

I see what you are saying -- msg are `asyncio.Future()` and the state will only be updated when we await that message.  @madsbk / @Akshay-Venkatesh is there another await for messages besides the call to `update_pending_messages`

Reply:
I don't know if there is.

For context, this part of the code needs extensive testing and not just the unit tests in tests directory. We'll need benchmarks and dask workloads also running before we can accept the change that come out of here